### PR TITLE
aws_kms - stabilize and add integration tests

### PR DESCRIPTION
**Pull request by jatorcasso**
##### SUMMARY
- update/add integration tests for various actions
- return list of policies as a list of jsons for clarity
- sleep on updates (no kms waiter, attempted manual waiters but still had test failures)

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
aws_kms

##### ADDITIONAL INFORMATION
I tried adding manual waiters for different actions like waiting for tags to be correct, policy to be updated, etc, but would still fail ~half of the time on idempotency tests. seems like after updating the key's status is a bit buggy.

### CODE REVIEW

File: tests/integration/targets/aws_kms/tasks/main.yml
Comment (jillr) [inline tests/integration/targets/aws_kms/tasks/main.yml:694]:
```suggestion
        key: '{{ info_tag_filtered.kms_keys[0] }}'
```

File: changelogs/fragments/1052-aws_kms-stabilize-integration-tests.yml
Comment (jillr) [inline changelogs/fragments/1052-aws_kms-stabilize-integration-tests.yml:2]: Nit, could you add some words explaining why we wait after update?  Something like, "for requested changes to be reflected on AWS" maybe?  Otherwise folks might be confused why the module takes longer to run updates now.

**Review by jillr — APPROVED** [review]
LGTM, there's a lot of formatting changes in the tests though (which are good changes!) so I'd like to see a second review before a merge in case I missed anything.

**Comment by jatorcasso** [comment]
Considering the last CI run failed, it looks like the wait will still cause CI problems. Seems like I should just remove the wait and keep the tests marked as `unstable` @jillr @alinabuzachis

File: tests/integration/targets/aws_kms/tasks/main.yml
Comment (tremble) [inline tests/integration/targets/aws_kms/tasks/main.yml:232]:
```suggestion
        policy: "{{ lookup('template', 'console-policy.j2') }}"
```

Is this actually a change in the policy?  At first glance, the check's failing because nothing changed...

Comment (jatorcasso) [reply tests/integration/targets/aws_kms/tasks/main.yml:232]: Later on in the tests a _different_ policy is updated. Hope that clears things up!

Comment (jatorcasso) [reply tests/integration/targets/aws_kms/tasks/main.yml:232]: well, the only reason why it ran twice is because the sleep(5) wasn't enough to wait for the tags to completely change. I tried different ways of waiting for tag update when I was messing with this before defaulting to sleep(), so I'm not really sure what else can be done here

**Review by mandar242 — APPROVED** [review]
LGTM!

**Review by markuman — APPROVED** [review]

File: plugins/modules/aws_kms.py
Comment (alinabuzachis) [inline plugins/modules/aws_kms.py:909]: I'm not a fan of using sleep here because it isn't probably solving the issue. If sometimes the update operation will take more than 5 seconds we will be in the same situation. Maybe @tremble can suggest a better approach instead.

File: plugins/modules/aws_kms.py
Comment (alinabuzachis) [inline plugins/modules/aws_kms.py:355]: Not sure for the name of this param, it is actually a list of dicts.

Comment (jatorcasso) [reply plugins/modules/aws_kms.py:355]: I guess it can be `policies_dicts` ? Im not sure. `policies` is the natural name.. or `policies_list` ?

Comment (jillr) [reply plugins/modules/aws_kms.py:355]: `kms_policies`?

File: plugins/modules/aws_kms.py
Comment (jillr) [reply plugins/modules/aws_kms.py:909]: I think what we'd need to do here is a build a comparison function that constructs what we expect to returned by `get_key_details()`, and checks to see if `result` matches (or at least the relevant keys).  If it doesn't, retry (some limited number of times).

Or else change every update/ensure function to do something similar - "connection.update_key_description, then connection.describe_key and check the description".  I don't like this option much though.

Comment (markuman) [reply plugins/modules/aws_kms.py:909]: > I think what we'd need to do here is a build a comparison function that constructs what we expect to returned by get_key_details(), and checks to see if result matches (or at least the relevant keys). If it doesn't, retry (some limited number of times).

`sleep`  is used like this in [s3_lifecycle](https://github.com/ansible-collections/community.aws/blob/main/plugins/modules/s3_lifecycle.py#L486-L494)

File: plugins/modules/aws_kms.py
Comment (jatorcasso) [reply plugins/modules/aws_kms.py:355]: turns out there can only be one policy attached. maybe `policy` makes the most sense here?

File: plugins/modules/aws_kms.py
Comment (jatorcasso) [reply plugins/modules/aws_kms.py:909]: @tremble - tagging you here because this is simliar to the igw issue where on one call of getting the key, description is _new-description_, then on following call, description is _old-description_. same with tags, policies, etc

Comment (jillr) [reply plugins/modules/aws_kms.py:909]: @jatorcasso Gotcha.  In that case the sleep sounds justified, can you please update the code comment to briefly explain that the API stays inconsistent too long to do a simple check and we need to wait to be confident the change has propagated?  Otherwise we might try to change it later and run into this problem again.

Comment (jatorcasso) [reply plugins/modules/aws_kms.py]: To summarize the OT (Slack) conversation around this - it seems that all of the proposed workarounds regarding the wait period after updates and inconsistencies in whats being reflected when fetching key metadata don't fix the problem (integration tests still failing on idempotency tests). Documenting this in the module as well as adding waits in integration tests was decided to be the best solution (albeit not a great one).

File: plugins/modules/aws_kms.py
Comment (jillr) [inline plugins/modules/aws_kms.py:320]:
```suggestion
  description: Whether the automatic annual key rotation is enabled. Returns None if key rotation status can't be determined.
```

**Review by jillr — APPROVED** [review]
thanks for all your work on this @jatorcasso!

**Pull request merged**
